reconcile/utils/acs_api.py

```diff
@@ -0,0 +1,384 @@
+        except ValueError as e:
+            # it is valid for the default Unrestricted access scope to have null `rules`
+            if api_data.get("name") != "Unrestricted":
+                raise e
+            unrestricted = True
+
+        super().__init__(
+            id=api_data["id"],
+            name=api_data["name"],
+            clusters=[]
+            if unrestricted
+            else api_data["rules"].get("includedClusters", []),
+            namespaces=[]
+            if unrestricted
+            else api_data["rules"].get("includedNamespaces", []),
+            description=api_data.get("description", ""),
+        )
+
+
+class PermissionSet(BaseModel):
+    id: str
+    name: str
+
+    def __init__(self, api_data: Any) -> None:
+        # attributes defined within stackrox(ACS) API for GET /v1/permissionsets/{id}
+        check_len_attributes(["id", "name"], api_data)
+
+        super().__init__(id=api_data["id"], name=api_data["name"])
+
+
+def check_len_attributes(attrs: list[Any], api_data: Any) -> None:
+    # generic attribute check function for expected types with valid len()
+    for attr in attrs:
+        value = api_data.get(attr)
+        if value is None or len(value) == 0:
+            raise ValueError(
+                f"Attribute '{attr}' must exist and not be empty\n\t{api_data}"
+            )
+
+
+class AcsApi:
+    def __init__(
+        self,
+        instance: Any,
+        timeout: int = 30,
+    ) -> None:
+        self.url = instance["url"]
+        self.token = instance["token"]
+        self.timeout = timeout
+
+    def generic_get_request(self, path: str) -> requests.Response:
+        response = requests.get(
+            url=f"{self.url}{path}",
+            headers={"Authorization": f"Bearer {self.token}"},
+            timeout=self.timeout,
+        )
+        try:
+            response.raise_for_status()
+        except requests.exceptions.RequestException as details:
+            raise requests.exceptions.RequestException(
+                f"Failed to perform GET request:\n\t{details}\n\t{response.text}"
+            )
+
+        return response
+
+    def generic_post_request(self, path: str, json: Any) -> requests.Response:
+        response = requests.post(
+            url=f"{self.url}{path}",
+            headers={
+                "Authorization": f"Bearer {self.token}",
+                "Content-Type": "application/json",
+            },
+            timeout=self.timeout,
+            json=json,
+        )
+
+        try:
+            response.raise_for_status()
+        except requests.exceptions.RequestException as details:
+            raise requests.exceptions.RequestException(
+                f"Failed to perform POST request with body:\n\t{json}\n\t{details}\n\t{response.text}"
+            )
+
+        return response
+
+    def generic_put_request(self, path: str, json: Any) -> requests.Response:
+        response = requests.put(
```

Comment:
Perhaps you could consolidate some of these request functions into a single function? Maybe like:
`def generic_request(self, path, request_type, json as an optional any)`

To cut down on duplicated code.

Reply:
Yep, good call. Kept appending request functions as I progressed through the reconcile operations and completely forgot to circle back on the sprawl of redundant code :sweat_smile: 

Pushed an update to consolidate